src/objective/regression_objective.hpp

```diff
@@ -51,13 +51,88 @@ class RegressionL2loss: public ObjectiveFunction {
   const float* weights_;
 };
 
-class RegressionLHuberLoss: public ObjectiveFunction {
+
+class RegressionL1loss: public ObjectiveFunction {
+public:
+  explicit RegressionL1loss(const ObjectiveConfig& config) {}
+
+  ~RegressionL1loss() {}
+
+  void Init(const Metadata& metadata, data_size_t num_data) override {
+    num_data_ = num_data;
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+  }
+
+  void GetGradients(const score_t* score, score_t* gradients,
+                    score_t* hessians) const override {
+    if (weights_ == nullptr) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+
+        if (diff >= 0.0) {
+          gradients[i] = 1.0;
+        } else {
+          gradients[i] = -1.0;
+        }
+
+        /*
+         * approximate hessians with Gaussian function
+         * cf. https://en.wikipedia.org/wiki/Gaussian_function
+         */
+        const double x = (std::fabs(diff) > 0.0) ? std::fabs(diff) : 1.0e-6;
+        const double a = 2.0;  // difference of two first derivatives, (zero to inf) and (zero to -inf)
+        const double b = 0.0;
+        const double c = (std::fabs(score[i]) + std::fabs(label_[i])) / 1.0e3;
+        hessians[i] = std::exp(-(x - b) * (x - b) / 2.0 * c * c) / a;
+      }
+    } else {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+
+        if (diff >= 0.0) {
+          gradients[i] = weights_[i];
+        } else {
+          gradients[i] = -weights_[i];
+        }
+
+        /*
+         * approximate hessians with Gaussian function
+         * cf. https://en.wikipedia.org/wiki/Gaussian_function
+         */
+        const double x = (std::fabs(diff) > 0.0) ? std::fabs(diff) : 1.0e-6;
+        const double a = 2.0 * weights_[i];  // difference of two first derivatives, (zero to inf) and (zero to -inf).
+        const double b = 0.0;
+        const double c = (std::fabs(score[i]) + std::fabs(label_[i])) / 1.0e3;
+        hessians[i] = std::exp(-(x - b) * (x - b) / 2.0 * c * c) / a;
```

Comment:
I think you can use an inline function to calculate hessian and avoid these repeated code.